### DISCUSSION TITLE
Why so slow on Windows?

### DISCUSSION THREAD

**Discussion by emiltin** [ruby/setup-ruby#635] [General]
Thank you for your excellent work.

I'm wondering why installing ruby and gems with ruby/setup-ruby@v1 is so much slower on windows? An example:

windows-latests:  2m18s
ubuntu-latest: 2s
macos-latest: 3s

Is this normal or is something wrong with my setup?

Thank you!
Emil

Windows log:
```
Run ruby/setup-ruby@v1
Using 3.3 as input from file .tool-versions
Downloading Ruby
Extracting  Ruby
Modifying Path
Downloading msys2 build tools
Extracting  msys2 build tools
Downloading ucrt64-3.0 build tools
Extracting  ucrt64-3.0 build tools
Adding ridk env variables
Print Ruby version
Installing Bundler
> bundle install
C:\Windows\system32\cmd.exe /D /S /C "C:\hostedtoolcache\windows\Ruby\3.3.4\x64\bin\bundle.bat config --local path D:\a\rsmp_validator\rsmp_validator\vendor\bundle"
C:\Windows\system32\cmd.exe /D /S /C "C:\hostedtoolcache\windows\Ruby\3.3.4\x64\bin\bundle.bat config --local deployment true"
Cache key: setup-ruby-bundler-cache-v6-windows-2022-x64-ruby-3.3.4-wd-D:\a\rsmp_validator\rsmp_validator-with--without--only--Gemfile.lock-34a4b837306640ff25ed9bd6adfd82d6fde859ca11796cb6671a103a4a3f9d98
Cache Size: ~6 MB (6034436 B)
"C:\Program Files\Git\usr\bin\tar.exe" -xf D:/a/_temp/9c9a84a6-07cf-40b9-b4cd-c820c04bfe79/cache.tzst -P -C D:/a/rsmp_validator/rsmp_validator --force-local --use-compress-program "zstd -d"
Cache restored successfully
Found cache for key: setup-ruby-bundler-cache-v6-windows-2022-x64-ruby-3.3.4-wd-D:\a\rsmp_validator\rsmp_validator-with--without--only--Gemfile.lock-34a4b837306640ff25ed9bd6adfd82d6fde859ca11796cb6671a103a4a3f9d98
Received 6034436 of 6034436 (100.0%), 5.8 MBs/sec
C:\Windows\system32\cmd.exe /D /S /C "C:\hostedtoolcache\windows\Ruby\3.3.4\x64\bin\bundle.bat install --jobs 4"
Bundle complete! 5 Gemfile dependencies, 37 gems now installed.
Bundled gems are installed into `./vendor/bundle`
Took   2.99 seconds
```

Linux log:
```
Run ruby/setup-ruby@v1
Using 3.3 as input from file .tool-versions
Modifying PATH
Downloading Ruby
Extracting  Ruby
Print Ruby version
Installing Bundler
> bundle install
/opt/hostedtoolcache/Ruby/3.3.5/x64/bin/bundle config --local path /home/runner/work/rsmp_validator/rsmp_validator/vendor/bundle
/opt/hostedtoolcache/Ruby/3.3.5/x64/bin/bundle config --local deployment true
Cache key: setup-ruby-bundler-cache-v6-ubuntu-22.04-x64-ruby-3.3.5-wd-/home/runner/work/rsmp_validator/rsmp_validator-with--without--only--Gemfile.lock-5bcd7657dc6362f9d317d1ff9ef2e1f587a16bcfa48e8b4184172a1be4758117
Cache Size: ~7 MB (7409391 B)
/usr/bin/tar -xf /home/runner/work/_temp/15a678f8-748b-4a0e-b925-d66006c7fc8f/cache.tzst -P -C /home/runner/work/rsmp_validator/rsmp_validator --use-compress-program unzstd
Cache restored successfully
Found cache for key: setup-ruby-bundler-cache-v6-ubuntu-22.04-x64-ruby-3.3.5-wd-/home/runner/work/rsmp_validator/rsmp_validator-with--without--only--Gemfile.lock-5bcd7657dc6362f9d317d1ff9ef2e1f587a16bcfa48e8b4184172a1be4758117
/opt/hostedtoolcache/Ruby/3.3.5/x64/bin/bundle install --jobs 4
Bundle complete! 5 Gemfile dependencies, 37 gems now installed.
Bundled gems are installed into `./vendor/bundle`
Took   0.83 seconds
```

**Comment by mvz** [comment]
Looking at the timing data, it seems the 'Extracting ..' steps are really slow.

**Comment by MSP-Greg** [comment]
A long time ago it was decided that setup-ruby would install the compiler, build tools, and packages needed to build Ruby on Windows.

Ubuntu & macOS images have all these items preinstalled, Windows images do not.

Many repos may lack the experience with Windows to install the tools (the tools vary by Ruby version), so by preinstalling them, a common comment is 'it just works'...

**Comment by eregon** [comment]
One possibility, if you don't need to compile any gem native extensions, is to use:
```yml
with:
  windows-toolchain: none
```
That should speed it up quite a bit.
Probably it's a trick we should mention in the README in the Windows section (PR welcome).

**Reply by mvz** [reply]
Any idea why the extracting steps themselves are so slow? Extracting Ruby took 28 seconds in a recent run. Do these steps do more than extracting the `.7z` file?

**Reply by eregon** [reply]
It's literally just a `7z` command: https://github.com/ruby/setup-ruby/blob/a2bbe5b1b236842c1cb7dd11e8e3b51e0a616acc/windows.js#L173-L175
I think the fundamental issue here is Windows filesystems like NFTS are very slow, especially when it's about creating many small files.
If someone finds a faster way to extract archives on Windows that'd be interesting.

**Comment by MSP-Greg** [comment]
@mvz

Are you using a GitHub runner (not self-hosted)?

I think I might have a fix.  Currently, GitHub runners have both a HDD and an SSD.  The Ruby 7z files are being extracted to the HDD for CRuby versions that aren't head builds.

Looking at a recent CI run, Ruby 3.3.6 took 26.0 seconds to extract to the HDD, and Ruby ucrt/head took 2.5 seconds to extract to the SSD.

**Reply by mvz** [reply]
Yes, I'm using a GitHub runner. A slow HDD would certainly explain the problem. I also found https://github.com/actions/setup-dotnet/issues/260#issuecomment-1790162905. Is the C: drive the HDD?

**Reply by MSP-Greg** [reply]
> Is the C: drive the HDD?

Yes. D: is the SSD, which is where the repo is cloned to.

**Reply by mvz** [reply]
Would it be possible to change the toolcache path on Windows as defined in https://github.com/ruby/setup-ruby/blob/a2bbe5b1b236842c1cb7dd11e8e3b51e0a616acc/common.js#L302 to the D: drive?

**Reply by eregon** [reply]
In that place no, it must match what GitHub runners set as the toolcache, so e.g. a Ruby already installed in the image is picked.
But I suppose if we find out there is no Ruby already installed we could place it somewhere else.
I won't have time to write that change or review it soon though.

**Reply by MSP-Greg** [reply]
Sorry for not responding earlier.  I changed this to extract the builds to the D: drive.  One problem related to the Windows binstubs issue.

I think some older versions of RubyGems/Bundler (on Windows) hardcode the Ruby exec path in a binstub that is not located within the Ruby exe file folder, which is normal with gems installed with Bundler.  I believe newer binstubs can use the Ruby selected by `PATH`.

The problem is that with caching of `bundle install`, the binstubs use the Ruby exec path on `C:`.  When we move Ruby to `D:`, the cached binstubs don't work.

Revising the cache key (invalidating **all** caches) would solve the problem.  IIRC, it's only an issue with Ruby 2.5 and earlier.

**Reply by MSP-Greg** [reply]
I looked locally, and some of the Ruby 2.5 binstubs have a hardcoded Ruby exec path. Others resolve to `PATH`.

**Reply by mvz** [reply]
Thanks for looking into this, @eregon and @MSP-Greg.

Would a solution similar to https://github.com/actions/setup-go/pull/393 work?
